docs/userguide/mpi_apps.rst

```diff
@@ -1,335 +1,132 @@
-                        walltime=user_opts["walltime"]
-                    ),
-                ),
-            ],
-    )
-
-
-
-Next we define the CosmicTagger MPI application. TODO: Ask Khalid for help.
+    @bash_app
+    def lammps_mpi_application(infile: File, parsl_resource_specification: Dict):
+        from subprocess import run
+        with open('stdout.lmp', 'w') as fp, open('stderr.lmp', 'w') as fe:
+            proc = run(['$PARSL_MPI_PREFIX', '-i', 'in.lmp'], stdout=fp, stderr=fe)
+            return proc.returncode
 
-.. code-block:: python
 
-    @parsl.bash_app
-    def cosmic_tagger(workdir: str,
-                      datatype: str = "float32",
-                      batchsize: int = 8,
-                      framework: str = "torch",
-                      iterations: int = 500,
-                      trial: int = 2,
-                      stdout=parsl.AUTO_LOGNAME,
-                      stderr=parsl.AUTO_LOGNAME,
-                      parsl_resource_specification:Dict={}):
-        NRANKS = parsl_resource_specification['num_ranks']
-
-        return f"""
-        module purge
-        module use /soft/modulefiles/
-        module load conda/2023-10-04
-        conda activate
-
-        echo "PARSL_MPI_PREFIX : $PARSL_MPI_PREFIX"
-
-        $PARSL_MPI_PREFIX --cpu-bind numa \
-            python {workdir}/bin/exec.py --config-name a21 \
-                run.id=run_plrs_ParslDemo_g${NRANKS}_{datatype}_b{batchsize}_{framework}_i{iterations}_T{trial} \
-                run.compute_mode=GPU \
-                run.distributed=True \
-                framework={framework} \
-                run.minibatch_size={batchsize} \
-                run.precision={datatype} \
-                mode.optimizer.loss_balance_scheme=light \
-                run.iterations={iterations}
-        """
-
-In this example, we run a simple test that does an exploration over the ``batchsize`` parameter
-while launching the application over 2-4 nodes.
+Run either App by calling with its arguments and a resource specification which defines how to execute it
 
 .. code-block:: python
 
-    def run_cosmic_tagger():
-        futures = {}
-        for num_nodes in [2, 4]:
-            for batchsize in [2, 4, 8]:
-
-                parsl_res_spec = {"num_nodes": num_nodes,
-                                  "num_tasks": num_nodes * 4,
-                                  "ranks_per_node": 4}
-                future = cosmic_tagger(workdir="/home/yadunand/CosmicTagger",
-                                       datatype="float32",
-                                       batchsize=str(batchsize),
-                                       parsl_resource_specification=parsl_res_spec)
-
-
-                print(f"Stdout : {future.stdout}")
-                print(f"Stderr : {future.stderr}")
-                futures[(num_nodes, batchsize)] = future
+    # Resources in terms of nodes and how ranks are to be distributed are set on a per app
```

Comment:
Do we _always_ need a `parsel_ressource_specification`? I'm just thinking back to the approach used in your older videos where we never used such a thing with the `HighThroughputExecutor`. If it's omitted, I assume the implicit assumption is that all tasks have the same resource requirements but it will still function?

Reply:
There has to be some default made in that case. In the workqueue case, for example, ommitting a core count (or memory specification) for a task will allocate one entire worker node (whatever the relevant worker node happens to have on it).

That is one obvious default. The other is that you get 1 rank.

Reply:
Yes, you do always need one and - at the moment (see #3427) - Parsl will crash without it being provided as a keyword argument.

Reply:
Makes sense. This is not your problem to deal with so to speak, but this just makes it even more awkward that `parsl_resource_specification` is not a keyword argument of the decorator itself. Having to redefine all of your functions to take this (hidden) keyword argument just to parallelize things is a bit hacky and counter-intuitive.